src/carma-1-tenth/calc_route_deviation.py

```diff
@@ -0,0 +1,123 @@
+        return None
+    return point_arr[min_index]
+
+def calc_route_deviation(bag_dir, start_offset=0.0):
+    metadatafile : str = os.path.join(bag_dir, "metadata.yaml")
+    if not os.path.isfile(metadatafile):
+        raise ValueError("Metadata file %s does not exist. Are you sure %s is a valid rosbag?" % (metadatafile, bag_dir))
+    with open(metadatafile, "r") as f:
+        metadata_dict : dict = yaml.load(f, Loader=yaml.SafeLoader)["rosbag2_bagfile_information"]
+    storage_id = metadata_dict['storage_identifier']
+    topic_types, type_map, topic_metadata_map, reader = open_bagfile(bag_dir, storage_id=storage_id)
+    odom_topic = '/amcl_pose'
+    route_topic = '/route_graph'
+    topic_count_dict = {entry["topic_metadata"]["name"] : entry["message_count"] for entry in metadata_dict["topics_with_message_count"]}
+    topic_counts = np.array( list(topic_count_dict.values()) ) 
+    total_msgs = np.sum( topic_counts )
+    msg_dict = {key : [] for key in topic_count_dict.keys()}
+    filt = rosbag2_py.StorageFilter([odom_topic, route_topic])
+    reader.set_filter(filt)
+    route_graph = None
+    odom_count = 0
+    odometry = np.zeros((topic_count_dict[odom_topic], 2))
+    for idx in tqdm.tqdm(iterable=range(topic_count_dict[odom_topic] + topic_count_dict[route_topic])):
+        if(reader.has_next()):
+            (topic, data, t_) = reader.read_next()
+            msg_type = type_map[topic]
+            msg_type_full = get_message(msg_type)
+            msg = deserialize_message(data, msg_type_full)
+            if msg_type == "visualization_msgs/msg/MarkerArray":
+                route_graph = msg
+            else:
+                odometry[odom_count] = [-msg.pose.pose.position.y, msg.pose.pose.position.x]
+                odom_count += 1
+    route_coordinates = []
+    for i in range(len(route_graph.markers)):
+        if route_graph.markers[i].type == 2:
+            route_coordinates.append([-route_graph.markers[i].pose.position.y, route_graph.markers[i].pose.position.x])
+    route_coordinates_with_distance = np.zeros((len(route_coordinates), 3))
+    route_coordinates_with_distance[:, 1:] = np.array(route_coordinates)
+    running_distance = 0.0
+    for i in range(len(route_coordinates)):
+        if i > 0:
+            running_distance += np.linalg.norm(route_coordinates_with_distance[i, 1:] - route_coordinates_with_distance[i-1, 1:])
+        route_coordinates_with_distance[i,0] = running_distance
+    x_spline = make_interp_spline(route_coordinates_with_distance[:,0], route_coordinates_with_distance[:,1])
+    y_spline = make_interp_spline(route_coordinates_with_distance[:,0], route_coordinates_with_distance[:,2])
+    samples = np.linspace(route_coordinates_with_distance[0,0], route_coordinates_with_distance[-1,0], 5000)
+    route_x_points = x_spline(samples)
+    route_y_points = y_spline(samples)
+    route_deviations = []
+    odom_movements = []
+    for i in range(1, len(odometry)):
```

Comment:
Seeing as there's a lot of sequential for loops here I'd split them up with comments that briefly say what each is doing

Reply:
Added some comments for each code block